pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1181,36 +1181,41 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta algorithm
 	return true, nil, nil
 }
 
-// podMatchesPodAffinityTerms checks if the "targetPod" matches the given "terms"
-// of the "pod" on the given "nodeInfo".Node(). It returns three values: 1) whether
-// targetPod matches all the terms and their topologies, 2) whether targetPod
-// matches all the terms label selector and namespaces (AKA term properties),
-// 3) any error.
-func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulercache.NodeInfo, terms []v1.PodAffinityTerm) (bool, bool, error) {
-	if len(terms) == 0 {
-		return false, false, fmt.Errorf("terms array is empty")
-	}
-	props, err := getAffinityTermProperties(pod, terms)
+// podMatchesPodAffinityTerms checks if the "targetPod" matches the "checkAffinity (true:affinity,false:anti-affinity)"
+// of the "pod" on the given "nodeInfo".Node(). It returns four values: 1) whether
+// targetPod matches any of the affnityTerms and their topologies, 2) whethertargetPod matches
+// all the affnityTerms and their topologies, 3) whether targetPodmatches all the terms label
+// selector and namespaces (AKA term properties), 4) any error.
+func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulercache.NodeInfo, checkAffinity bool) (bool, bool, bool, error) {
+	matchedTerms, affinityTermsNum, err := getMatchingAffinityTerms(pod, targetPod, checkAffinity)
 	if err != nil {
-		return false, false, err
+		return false, false, false, err
 	}
-	if !podMatchesAllAffinityTermProperties(targetPod, props) {
-		return false, false, nil
-	}
-	// Namespace and selector of the terms have matched. Now we check topology of the terms.
 	targetPodNode, err := c.info.GetNodeInfo(targetPod.Spec.NodeName)
 	if err != nil {
-		return false, false, err
+		return false, false, false, err
 	}
-	for _, term := range terms {
+	if len(matchedTerms) == 0 {
```

Comment:
when `checkAffinity` is true and `len(matchedTerms) != affinityTermsNum` you should return `false, nil`.